File: python/01_core/03_properties/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
PySide6 属性系统示例

属性系统允许在运行时查询和修改对象属性：
- @Property 装饰器声明属性
- 支持通知信号 (notify)
- 支持动态属性

注意：
- PySide6的Property装饰器与C++ Q_PROPERTY类似
- PySide6没有Qt6 C++的QProperty绑定系统
- 在Python中可以用property + 观察者模式实现绑定效果

官方文档: https://doc.qt.io/qtforpython/PySide6/QtCore/Property.html
"""

import sys
from typing import Any
from PySide6.QtCore import (
    QCoreApplication,
    QObject,
    Property,
    Signal,
    Slot
)


class Rectangle(QObject):
    """
    矩形类 - 展示基础属性
    """
    
    # 信号定义
    widthChanged = Signal()
    heightChanged = Signal()
    areaChanged = Signal()
    perimeterChanged = Signal()
    
    def __init__(self, parent=None):
        super().__init__(parent)
        self._width = 0.0
        self._height = 0.0
    
    # Width 属性
    @Property(float, notify=widthChanged)
    def width(self):
        return self._width
    
    @width.setter
    def width(self, w):
        # 使用小容差比较浮点数，避免精度问题
        if abs(self._width - w) > 1e-10:
            self._width = w
            self.widthChanged.emit()
            self.areaChanged.emit()
            self.perimeterChanged.emit()
    
    # Height 属性
    @Property(float, notify=heightChanged)
    def height(self):
        return self._height
    
    @height.setter
    def height(self, h):
        if abs(self._height - h) > 1e-10:
            self._height = h
            self.heightChanged.emit()
            self.areaChanged.emit()
            self.perimeterChanged.emit()
    
    # 只读计算属性 - 面积
    @Property(float, notify=areaChanged)
    def area(self):
        return self._width * self._height
    
    # 只读计算属性 - 周长
    @Property(float, notify=perimeterChanged)
    def perimeter(self):
        return 2 * (self._width + self._height)
    
    # 常量属性 - 没有notify信号
    @Property(str, constant=True)
    def type(self):
        return "Rectangle"


class BindableRectangle(QObject):
    """
    可绑定矩形类
    注意：PySide6没有Qt6 C++的QProperty绑定系统
    这里使用Python的property和观察者模式实现类似效果
    """
    
    widthChanged = Signal()
    heightChanged = Signal()
    areaChanged = Signal()
    
    def __init__(self, parent=None):
        super().__init__(parent)
        self._width = 0.0
        self._height = 0.0
        self._area_bindings = []  # 绑定回调列表
        
        # 连接信号以实现自动计算
        self.widthChanged.connect(self._update_area)
        self.heightChanged.connect(self._update_area)
    
    @Property(float, notify=widthChanged)
    def width(self):
        return self._width
    
    @width.setter
    def width(self, w):
        if abs(self._width - w) > 1e-10:
            self._width = w
            self.widthChanged.emit()
    
    @Property(float, notify=heightChanged)
    def height(self):
        return self._height
    
    @height.setter
    def height(self, h):
        if abs(self._height - h) > 1e-10:
            self._height = h
            self.heightChanged.emit()
    
    @Property(float, notify=areaChanged)
    def area(self):
        """面积 - 自动计算"""
        return self._width * self._height
    
    def _update_area(self):
        """更新面积并通知绑定"""
        self.areaChanged.emit()
        # 调用所有绑定的回调
        for callback in self._area_bindings:
            callback(self.area)
    
    def subscribe_area(self, callback):
        """
        订阅面积变化（模拟Qt6的bindable property subscribe）
        """
        self._area_bindings.append(callback)


def demonstrate_dynamic_typing():
    """
    展示Python的动态类型特性
    对应C++的QVariant示例
    """
    print("\n=== Python 动态类型示例 ===")
    
    # Python是动态类型，变量可以存储任意类型
    v1 = 42
    v2 = "Hello"
    v3 = 3.14
    v4 = ["a", "b", "c"]
    
    print(f"v1 (int): {v1}, 类型: {type(v1).__name__}")
    print(f"v2 (str): {v2}, 类型: {type(v2).__name__}")
    print(f"v3 (float): {v3}, 类型: {type(v3).__name__}")
    print(f"v4 (list): {v4}, 类型: {type(v4).__name__}")
    
    # 类型检查
    print("\n类型检查:")
    print(f"v1 是 int 类型: {isinstance(v1, int)}")
    print(f"v2 是 int 类型: {isinstance(v2, int)}")
    
    # 类型转换
    num_str = "123"
    print(f"\n'123' 转 int: {int(num_str)}")
    
    # Python的Any类型（类似QVariant的概念）
    def print_any(value: Any):
        print(f"任意类型值: {value}, 类型: {type(value).__name__}")
    
    print("\n使用Any类型:")
    print_any(42)
    print_any("Hello")
    print_any([1, 2, 3])


def main():
    """主函数"""
    app = QCoreApplication(sys.argv)
    
    print("=== PySide6 属性系统示例 ===\n")
    
    # ============ 基础属性 ============
    print("--- 基础属性演示 ---")
    rect = Rectangle()
    
    # 连接信号
    rect.areaChanged.connect(lambda: print(f"面积变化: {rect.area}"))
    
    # 使用 setter
    rect.width = 10
    rect.height = 5
    
    print(f"宽度: {rect.width}")
    print(f"高度: {rect.height}")
    print(f"面积: {rect.area}")
    print(f"周长: {rect.perimeter}")
    print(f"类型: {rect.type}")
    
    # ============ 通过属性系统访问 ============
    print("\n--- 通过 setProperty/property 访问 ---")
    rect.setProperty("width", 20)
    rect.setProperty("height", 10)
    
    print(f"width 属性: {rect.property('width')}")
    print(f"height 属性: {rect.property('height')}")
    print(f"area 属性: {rect.property('area')}")
    
    # ============ 可绑定属性 ============
    print("\n--- 可绑定属性演示 (Python实现) ---")
    brect = BindableRectangle()
    
    # 订阅面积变化
    def on_area_changed(area):
        print(f"绑定属性 - 面积自动更新为: {area}")
    
    brect.subscribe_area(on_area_changed)
    brect.areaChanged.connect(lambda: on_area_changed(brect.area))
    
    brect.width = 5
    brect.height = 4
    
    print(f"绑定矩形面积: {brect.area}")
    
    # 修改宽度, area 自动更新
    brect.width = 10
    
    # ============ Python动态类型 ============
    demonstrate_dynamic_typing()
    
    # ============ 列出所有属性 ============
    print("\n--- 枚举所有属性 ---")
    meta = rect.metaObject()
    for i in range(meta.propertyCount()):
        prop = meta.property(i)
        value = rect.property(prop.name())
        print(f"{prop.name()} = {value}")
    
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: cpp/01_core/03_properties/main.cpp
/**
 * Qt6 属性系统示例
 *
 * 属性系统允许在运行时查询和修改对象属性：
 * - Q_PROPERTY 宏声明属性
 * - 支持绑定、通知、设计器集成
 * - Qt6 引入了属性绑定系统
 *
 * Q_PROPERTY 语法:
 * Q_PROPERTY(type name
 *     READ getter
 *     [WRITE setter]
 *     [NOTIFY signal]
 *     [BINDABLE bindable]
 *     [RESET reset]
 *     [DESIGNABLE bool]
 *     [STORED bool]
 *     [CONSTANT]
 *     [FINAL])
 *
 * 官方文档: https://doc.qt.io/qt-6/properties.html
 */

#include <QCoreApplication>
#include <QObject>
#include <QProperty>
#include <QDebug>
#include <QVariant>
#include <QMetaProperty>

class Rectangle : public QObject
{
    Q_OBJECT

    // 基础属性
    Q_PROPERTY(qreal width READ width WRITE setWidth NOTIFY widthChanged)
    Q_PROPERTY(qreal height READ height WRITE setHeight NOTIFY heightChanged)

    // 只读计算属性
    Q_PROPERTY(qreal area READ area NOTIFY areaChanged)
    Q_PROPERTY(qreal perimeter READ perimeter NOTIFY perimeterChanged)

    // 常量属性
    Q_PROPERTY(QString type READ type CONSTANT)

public:
    explicit Rectangle(QObject *parent = nullptr)
        : QObject(parent), m_width(0), m_height(0) {}

    // Width
    qreal width() const { return m_width; }
    void setWidth(qreal w) {
        if (!qFuzzyCompare(m_width, w)) {
            m_width = w;
            emit widthChanged();
            emit areaChanged();
            emit perimeterChanged();
        }
    }

    // Height
    qreal height() const { return m_height; }
    void setHeight(qreal h) {
        if (!qFuzzyCompare(m_height, h)) {
            m_height = h;
            emit heightChanged();
            emit areaChanged();
            emit perimeterChanged();
        }
    }

    // 计算属性
    qreal area() const { return m_width * m_height; }
    qreal perimeter() const { return 2 * (m_width + m_height); }

    // 常量属性
    QString type() const { return "Rectangle"; }

signals:
    void widthChanged();
    void heightChanged();
    void areaChanged();
    void perimeterChanged();

private:
    qreal m_width;
    qreal m_height;
};

// Qt6 QProperty 绑定示例
class BindableRectangle : public QObject
{
    Q_OBJECT

    // Qt6 可绑定属性
    Q_PROPERTY(qreal width READ width WRITE setWidth BINDABLE bindableWidth)
    Q_PROPERTY(qreal height READ height WRITE setHeight BINDABLE bindableHeight)
    Q_PROPERTY(qreal area READ area BINDABLE bindableArea)

public:
    explicit BindableRectangle(QObject *parent = nullptr)
        : QObject(parent)
    {
        // 设置 area 的绑定表达式
        m_area.setBinding([this]() {
            return m_width.value() * m_height.value();
        });
    }

    qreal width() const { return m_width.value(); }
    void setWidth(qreal w) { m_width.setValue(w); }
    QBindable<qreal> bindableWidth() { return &m_width; }

    qreal height() const { return m_height.value(); }
    void setHeight(qreal h) { m_height.setValue(h); }
    QBindable<qreal> bindableHeight() { return &m_height; }

    qreal area() const { return m_area.value(); }
    QBindable<qreal> bindableArea() { return &m_area; }

private:
    Q_OBJECT_BINDABLE_PROPERTY(BindableRectangle, qreal, m_width)
    Q_OBJECT_BINDABLE_PROPERTY(BindableRectangle, qreal, m_height)
    Q_OBJECT_BINDABLE_PROPERTY(BindableRectangle, qreal, m_area)
};

void demonstrateQVariant()
{
    qDebug() << "\n=== QVariant 示例 ===";

    // QVariant 可以存储任意类型
    QVariant v1 = 42;
    QVariant v2 = "Hello";
    QVariant v3 = 3.14;
    QVariant v4 = QStringList{"a", "b", "c"};

    qDebug() << "v1 (int):" << v1.toInt() << "类型:" << v1.typeName();
    qDebug() << "v2 (string):" << v2.toString() << "类型:" << v2.typeName();
    qDebug() << "v3 (double):" << v3.toDouble() << "类型:" << v3.typeName();
    qDebug() << "v4 (list):" << v4.toStringList() << "类型:" << v4.typeName();

    // 类型检查
    qDebug() << "\n类型检查:";
    qDebug() << "v1 可转为 int:" << v1.canConvert<int>();
    qDebug() << "v2 可转为 int:" << v2.canConvert<int>();

    // 类型转换
    QVariant numStr = "123";
    qDebug() << "\n\"123\" 转 int:" << numStr.toInt();
}

int main(int argc, char *argv[])
{
    QCoreApplication app(argc, argv);

    qDebug() << "=== Qt6 属性系统示例 ===\n";

    // ============ 基础属性 ============
    qDebug() << "--- 基础属性演示 ---";
    Rectangle rect;

    // 连接信号
    QObject::connect(&rect, &Rectangle::areaChanged, [&rect]() {
        qDebug() << "面积变化:" << rect.area();
    });

    // 使用 setter
    rect.setWidth(10);
    rect.setHeight(5);

    qDebug() << "宽度:" << rect.width();
    qDebug() << "高度:" << rect.height();
    qDebug() << "面积:" << rect.area();
    qDebug() << "周长:" << rect.perimeter();
    qDebug() << "类型:" << rect.type();

    // ============ 通过属性系统访问 ============
    qDebug() << "\n--- 通过 setProperty/property 访问 ---";
    rect.setProperty("width", 20);
    rect.setProperty("height", 10);

    qDebug() << "width 属性:" << rect.property("width").toDouble();
    qDebug() << "height 属性:" << rect.property("height").toDouble();
    qDebug() << "area 属性:" << rect.property("area").toDouble();

    // ============ Qt6 可绑定属性 ============
    qDebug() << "\n--- Qt6 可绑定属性演示 ---";
    BindableRectangle brect;

    // 属性改变时自动通知
    brect.bindableArea().subscribe([&brect]() {
        qDebug() << "绑定属性 - 面积自动更新为:" << brect.area();
    });

    brect.setWidth(5);
    brect.setHeight(4);

    qDebug() << "绑定矩形面积:" << brect.area();

    // 修改宽度, area 自动更新
    brect.setWidth(10);

    // ============ QVariant ============
    demonstrateQVariant();

    // ============ 列出所有属性 ============
    qDebug() << "\n--- 枚举所有属性 ---";
    const QMetaObject *meta = rect.metaObject();
    for (int i = 0; i < meta->propertyCount(); ++i) {
        QMetaProperty prop = meta->property(i);
        qDebug() << prop.name() << "=" << rect.property(prop.name());
    }

    return 0;
}

#include "main.moc"
